docs/design/2022-06-15-phyical_plan_explain.md

```diff
@@ -0,0 +1,472 @@
+// many other node.
++- ...
+```
+
+What we are interested in here is the content in `CoprocessorRDD`. This node is the output of TiSpark, other nodes are the output of Spark. The content in the `CoprocessRDD` is the output of the `toStringInternal` method called by `TiDAGRequest`.
+
+First of all, let's introduce a few member variables in `TiDAGRequest`.
+
+* `indexInfo` is the index to be scanned by the physical plan, if it is empty, it means scanning the table.
+* `isDoubleRead` is `true` means scanning table after scanning the index, `false` means only scanning the index or scanning the table.
+* `filters` is the selection expression in the physical plan in normal execution.
+* `downgradeFilters` is the selection condition of the physical plan after the downgrade is triggered (what is downgrade will be described later).
+* `pushdownFilters` is only used in `toStringInternal`. When `indexInfo==null` or `isDoubleRead==false` or all the columns involved in the filter are in the index, the content of `pushDownFilters` is equal to filters, otherwise, `pushDownFilters` is empty.
+* `keyRange` is the scan range of the table or index.
+
+The main code of `toStringInternal` is shown below.
+
+```java
+private String toStringInternal(){
+	StringBuilder sb=new StringBuilder();
+    // init some data in TiDAGRequest.
+	init();
+    switch (getIndexScanType()) {
+      case INDEX_SCAN:
+        sb.append("IndexScan");
+        isIndexScan=true;
+        break;
+      case COVERING_INDEX_SCAN:
+        sb.append("CoveringIndexScan");
+        break;
+      case TABLE_SCAN:
+        sb.append("TableScan");
+    }
+    if (isIndexScan && !getDowngradeFilters().isEmpty()) {
+       sb.append(", Downgrade Filter: ");
+       Joiner.on(", ").skipNulls().appendTo(sb, request.getDowngradeFilters());
+    }
+    if (!isIndexScan && !getFilters().isEmpty()) {
+        sb.append(", Residual Filter: ");
+        Joiner.on(", ").skipNulls().appendTo(sb, request.getFilters());
+    }
+    if (!getPushDownFilters().isEmpty()) {
+        sb.append(", PushDown Filter: ");
+        Joiner.on(", ").skipNulls().appendTo(sb, request.getPushDownFilters());
+    }
+    // add keyRange and other push down expression to sb.
+    ...
+    return sb.toString();
+}
+```
+
+From the code we can see that the execution plan is divided into three kinds of `IndexScan`, `CoveringIndexScan`, and `TableScan`; `TiDAGRequest` has three kinds `Filter` —— `Downgrade Filter`, ` Resdiual Filter`, `PushDown Filter`. There is zero or more `Filter` for an execution plan. We will describe the meaning of each execution plan and `Filter` in detail below.
+
+### Meaning of execution plan
+
+#### `IndexScan`
+
+In TiSpark, an `IndexScan` requires two scans. The first scan is scanning in index data and the second scan is scanning in table data to get the data we need from the results returned in the first scan.
```

Comment:
Could you please explain "an `IndexScan` requires two scans"? Does it mean that there are two steps(index scan and table scan) in one IndexScan? Or It means If we do common IndexScan (Non-CoveringIndexScan), we should do TableScan which is another scan-type after.

Reply:
It`s two steps. change the description.